File: perl/forking_primes/prime.pl
#!/usr/bin/perl

use strict;
use warnings;

my $min=$ARGV[0];
my $max=$ARGV[1];

sub is_prime{
	my $n=$_[0];
	my $prim=1;
	#proof that it works!
	#while(1){};
        for (my $j=2;$j<=sqrt($n);$j++){
                if ($n%$j==0){
                        $prim=0;
                        last;
                }
        }
        if ($prim==1){
                print "$n\n";
        }
}

my $parallel = 10;
my $running = 0;
my $result;

#$SIG{CHLD}="IGNORE";

for (my $i=$min;$i<=$max;$i++){
	my $pid=fork;
	if ($pid==0){
		&is_prime($i);
		exit;
	}
	else{
#		print "Running childs: ".($running+1)."\n";
                $running++;
                if ($running >= $parallel) {
#			print "Waiting...\n";
                        $result = wait;
                        $running--;
                }
	}
}

#wait for all children to finish
while (wait() != -1) {}
print "Done\n";

File: ruby/forking_primes/prime.rb
#!/usr/bin/ruby -w

min=ARGV.shift.to_i;
max=ARGV.shift.to_i;

def is_prime(n)
	@n=n
	prim=1
        (2..Math.sqrt(@n)).each do |j|
                if (@n%j==0) then
                        prim=0
                        break
                end
        end
        if (prim==1) then
                puts @n
        end
end

parallel = 5
running = 0

(min..max).each do |i|
	pid=Process.fork
	if (pid==nil) then
		is_prime(i)
		exit
	else
#		puts "Running children: "+(running+1).to_s
		running+=1
		if (running==parallel) then
			result=Process.wait()
			running-=1
		end
	end
end

while (Process.wait()!=-1) do end
puts "Done!"

